content/post/graceful-k8s-delpoyments/index.md

```diff
@@ -0,0 +1,209 @@
+
+Secondly it suffers from low spatial cohesion - the delay timing is now encoded in the application source code, while the upper bound (`spec.terminationGracePeriodSeconds`) is in the k8s object description. This makes the approach less readably (or understandable) because you have to look at 2 different files, possibly in 2 different code repositories, to understand the full picture. It also makes it harder to maintain, because you have to orchestrate application changes (delay timing) with configuration changes (`spec.terminationGracePeriodSeconds`).
+
+## Idea #3: _Dynamically_ delay app shutdown
+
+In internal discussions we outlined a 3rd approach that seems more robust than the others, but complicates the application:
+
+1. Run sure pod health checks every second
+1. Don’t sleep on pre-stop hook
+1. When the pod receives a `TERM` signal, start a stopwatch (on app layer).
+1. Reset the stopwatch to zero every time the pod receives a request (it means we’re still getting traffic and we should be getting health check traffic as).
+1. When the stopwatch reaches `N` seconds, in other words `N` seconds without traffic, **AND** all connections are drained, we can kill the pod.
+
+This may or may not work, depending on when k8s decides to stop sending health checks. We did not try this approach because it is a lot more complicated than a sleep and we can slower deployments. But _in theory_ it is more reliable than idea #1 and #2.
+
+## Conclusion
+
+In summary, we described the Kubernetes pod termination process, give recommendations on how to cleanly handle the `TERM` signal in different scenarios. Finally we describe several techniques that allow us to avoid failures during Kubernetes pod eviction, e.g. during rolling deployments. None of the techniques are perfect, idea #1 & #2 require magic numbers which rely on the latencies in your infrastructure, your application shutdown and the long-(/short)-lived-ness of connections. But we have successfully used these mitigation in production services with confirmed 0% error rate during deployments. Idea #3 is just that - an idea, at this point we have not implemented it, so your milage may vary.
+
+This is all based on Kubernetes clusters using kube-proxy & iptables. Kubernetes is moving towards [CNI implementations that use eBPF instead of iptables](https://archive.fosdem.org/2020/schedule/event/replacing_iptables_with_ebpf/) like [Cilium](https://cilium.io/) or [Calico](https://docs.projectcalico.org/about/about-calico). But the behavior described here also applies to those eBPF based solutions (or at least seems to apply to it) and there may be more caveats that you have to keep in mind. For instance, Cilium seems to have a [bug](https://github.com/cilium/cilium/issues/14844) that causes existing connections to also fail during pod termination, making graceful shutdown harder.
```

Comment:
Can we use footnote syntax here?

Reply:
Which part exactly do you want to put in a footnote? The whole paragraph, or just the last sentence referencing the cilium bug?

Reply:
ok, on second read it actually fits better into a footnote, I moved it.